File: src/shared/utils/hal.ts
/**
 * HAL href helpers shared by both trees.
 *
 * OpenProject links resources by href, not by id: a time entry carries its
 * work package as `_links.workPackage.href = "/api/v3/work_packages/12345"`.
 * The backend needs the id to *build* those hrefs, and the frontend needs
 * it to prefill the edit form from an existing entry — so the parsing lives
 * here rather than in either tree, exactly as `parseHoursToDecimal` does in
 * `./time.ts`. The backend has its own copy in `src-tauri/src/util/hal.rs`,
 * which is the one that decides anything — see `docs/architecture.md`.
 *
 * Nothing here trusts its input: hrefs are server-supplied, and the ids they
 * yield are fed straight back into request paths, so a segment that isn't a
 * positive integer yields `null` and the caller skips the resource. See
 * `docs/security.md`.
 */

/**
 * API path of the `TimeEntriesActivity` resource collection. The href → id
 * parsers below and the client's href builders both derive from these
 * constants so they can never drift apart.
 */
export const TIME_ENTRY_ACTIVITY_PATH = '/api/v3/time_entries/activities'

/** API path of the work package collection. */
export const WORK_PACKAGE_PATH = '/api/v3/work_packages'

/**
 * API path of the time entry collection — the base for the update and delete
 * request URLs, which append a validated numeric id.
 */
export const TIME_ENTRY_PATH = '/api/v3/time_entries'

/**
 * The collections a work package's editable `_links` point into.
 *
 * Both directions run through these constants: the parsers below read an id
 * out of a server-supplied href, and `buildWorkPackagePatchPayload` writes a
 * fresh href from a validated numeric id. Sharing the constant is what keeps
 * "the id we read" and "the href we send" from drifting apart.
 */
export const STATUS_PATH = '/api/v3/statuses'
export const TYPE_PATH = '/api/v3/types'
export const PRIORITY_PATH = '/api/v3/priorities'
export const PROJECT_PATH = '/api/v3/projects'
export const USER_PATH = '/api/v3/users'
export const GROUP_PATH = '/api/v3/groups'
export const PLACEHOLDER_USER_PATH = '/api/v3/placeholder_users'

/**
 * A HAL Formattable (`description`, `comment`), in whichever of its three
 * spellings arrived: the `{format, raw, html}` object a current instance sends,
 * a bare string on older ones, or `null` for an empty value.
 *
 * Structural, and duplicated in the bridge contract as `Formattable`: the
 * backend's `Formattable` enum in `src-tauri/src/schemas/common.rs` is what
 * produces the value, and this is the reader both trees use.
 */
export type HalFormattable =
  | { format?: string; raw?: string | null; html?: string | null }
  | string
  | null

/**
 * The raw text of a Formattable.
 *
 * Lives here so the edit draft can be seeded from a work package's description
 * without the frontend re-deriving "which of the three shapes is this" at every
 * call site — the same reason `parseHoursToDecimal` is shared rather than
 * duplicated. `html` is deliberately never read: it is the server's rendering
 * of `raw`, and the editor writes `raw`.
 */
export function formattableRaw(value: HalFormattable | undefined): string {
  if (value === undefined || value === null) return ''
  if (typeof value === 'string') return value
  return value.raw ?? ''
}

/**
 * Parse a resource id out of a HAL href under `collectionPath`.
 *
 * The trailing segment is validated as a positive integer rather than
 * trusted — a non-numeric, negative, or wrong-collection href yields `null`.
 * A leading origin (`https://host/api/v3/…`) is tolerated: the match is
 * anchored on the collection path and the end of the string, not on the start.
 */
export function parseResourceIdFromHref(
  collectionPath: string,
  href: unknown
): number | null {
  if (typeof href !== 'string') return null
  const match = new RegExp(`${collectionPath}/(\\d+)/?$`).exec(href)
  if (!match) return null
  const id = Number(match[1])
  return Number.isInteger(id) && id > 0 ? id : null
}

/** Parse a `TimeEntriesActivity` id out of its self href. */
export function parseActivityIdFromHref(href: unknown): number | null {
  return parseResourceIdFromHref(TIME_ENTRY_ACTIVITY_PATH, href)
}

/**
 * Parse a work package id out of a `_links.workPackage.href`.
 *
 * Used to prefill the edit form from an existing entry: the entry carries its
 * work package only as an href, while the update input takes a numeric id. An
 * href that yields nothing makes the entry non-editable, rather than producing
 * a request the server would reject.
 *
 * Note the two parsers are anchored on different collections — an activity
 * href (`/api/v3/time_entries/activities/3`) is not a work package href, and
 * neither parser accepts the other's input.
 */
export function parseWorkPackageIdFromHref(href: unknown): number | null {
  return parseResourceIdFromHref(WORK_PACKAGE_PATH, href)
}

/**
 * Parsers for the links the work-package editor reads.
 *
 * Each is anchored on its own collection, so a status href is never mistaken
 * for a type href — the ids overlap freely across collections, and a
 * cross-collection match would silently produce a valid-looking but wrong
 * request.
 */
export function parseStatusIdFromHref(href: unknown): number | null {
  return parseResourceIdFromHref(STATUS_PATH, href)
}

export function parseTypeIdFromHref(href: unknown): number | null {
  return parseResourceIdFromHref(TYPE_PATH, href)
}

export function parsePriorityIdFromHref(href: unknown): number | null {
  return parseResourceIdFromHref(PRIORITY_PATH, href)
}

/**
 * Parse a project id out of a `_links.project.href`.
 *
 * The frontend needs it because the assignee options come from a *project*
 * resource (`/api/v3/projects/{id}/available_assignees`) — see PLAN.md,
 * "Verified API shapes". A number is what crosses IPC; the backend
 * rebuilds the path from it.
 */
export function parseProjectIdFromHref(href: unknown): number | null {
  return parseResourceIdFromHref(PROJECT_PATH, href)
}

/**
 * Parse a principal id out of an assignee href.
 *
 * A principal is a user, a group, or a placeholder user, and OpenProject links
 * each from its own collection — so all three are tried. Only the id is
 * returned: which collection it came from is deliberately dropped, because the
 * editor only ever writes a `/api/v3/users/{id}` href back (assigning a group
 * is out of scope for this stage).
 */
export function parsePrincipalIdFromHref(href: unknown): number | null {
  for (const path of [USER_PATH, GROUP_PATH, PLACEHOLDER_USER_PATH]) {
    const id = parseResourceIdFromHref(path, href)
    if (id !== null) return id
  }
  return null
}

File: src-tauri/src/util/hal.rs
//! HAL collection paths and href readers.
//!
//! Port of `src/shared/utils/hal.ts` from the Electron app. OpenProject speaks
//! HAL+JSON: a resource reference is an href like `/api/v3/statuses/7`, and the
//! id has to be read back out of it. Every path constant lives here so a
//! request URL is never assembled from a string literal at the call site, and
//! so `parse_resource_id_from_href` can be *anchored* — a `status` href sitting
//! in the `type` slot yields `None` rather than a plausible-looking wrong id.
//!
//! The frontend keeps its own TypeScript copy of this module (it reads hrefs off
//! validated resources to seed forms); the two are deliberately parallel, and
//! both are covered by tests.

use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashMap;
use std::sync::Mutex;

pub const TIME_ENTRY_ACTIVITY_PATH: &str = "/api/v3/time_entries/activities";
pub const TIME_ENTRY_PATH: &str = "/api/v3/time_entries";
pub const WORK_PACKAGE_PATH: &str = "/api/v3/work_packages";
pub const STATUS_PATH: &str = "/api/v3/statuses";
pub const TYPE_PATH: &str = "/api/v3/types";
pub const PRIORITY_PATH: &str = "/api/v3/priorities";
pub const PROJECT_PATH: &str = "/api/v3/projects";
pub const USER_PATH: &str = "/api/v3/users";
pub const GROUP_PATH: &str = "/api/v3/groups";
pub const PLACEHOLDER_USER_PATH: &str = "/api/v3/placeholder_users";
pub const ATTACHMENT_PATH: &str = "/api/v3/attachments";

/// Compiled `<collection>/<digits>` matchers, one per collection path.
///
/// The set of paths is closed (the constants above), so the cache can never
/// grow unbounded from server input.
static ID_MATCHERS: Lazy<Mutex<HashMap<&'static str, Regex>>> =
    Lazy::new(|| Mutex::new(HashMap::new()));

fn id_matcher(collection_path: &'static str) -> Regex {
    let mut cache = ID_MATCHERS.lock().expect("href matcher cache poisoned");
    cache
        .entry(collection_path)
        .or_insert_with(|| {
            Regex::new(&format!("{}/(\\d+)/?$", regex::escape(collection_path)))
                .expect("collection path is a valid regex literal")
        })
        .clone()
}

/// Read the resource id out of an href, anchored on the collection it must
/// belong to.
///
/// `None` for anything that is not `<collection_path>/<positive integer>` —
/// including a null href, which is how HAL spells "unset".
pub fn parse_resource_id_from_href(
    collection_path: &'static str,
    href: Option<&str>,
) -> Option<i64> {
    let href = href?;
    let captures = id_matcher(collection_path).captures(href)?;
    let id: i64 = captures.get(1)?.as_str().parse().ok()?;
    if id > 0 {
        Some(id)
    } else {
        None
    }
}

pub fn parse_activity_id_from_href(href: Option<&str>) -> Option<i64> {
    parse_resource_id_from_href(TIME_ENTRY_ACTIVITY_PATH, href)
}

pub fn parse_work_package_id_from_href(href: Option<&str>) -> Option<i64> {
    parse_resource_id_from_href(WORK_PACKAGE_PATH, href)
}

/// A principal may be a user, a group, or a placeholder user — three
/// collections, one id. Tried in that order.
pub fn parse_principal_id_from_href(href: Option<&str>) -> Option<i64> {
    [USER_PATH, GROUP_PATH, PLACEHOLDER_USER_PATH]
        .into_iter()
        .find_map(|path| parse_resource_id_from_href(path, href))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reads_an_id_out_of_a_matching_href() {
        assert_eq!(
            parse_resource_id_from_href(STATUS_PATH, Some("/api/v3/statuses/7")),
            Some(7)
        );
        assert_eq!(
            parse_resource_id_from_href(STATUS_PATH, Some("/api/v3/statuses/7/")),
            Some(7)
        );
    }

    #[test]
    fn is_anchored_on_its_own_collection() {
        // The ids overlap freely across collections, so a type href in the
        // status slot must not yield a plausible-looking wrong id.
        assert_eq!(
            parse_resource_id_from_href(STATUS_PATH, Some("/api/v3/types/7")),
            None
        );
    }

    #[test]
    fn rejects_non_ids_and_unset_links() {
        assert_eq!(parse_resource_id_from_href(STATUS_PATH, None), None);
        assert_eq!(
            parse_resource_id_from_href(STATUS_PATH, Some("/api/v3/statuses/abc")),
            None
        );
        assert_eq!(
            parse_resource_id_from_href(STATUS_PATH, Some("/api/v3/statuses/0")),
            None
        );
        assert_eq!(
            parse_resource_id_from_href(STATUS_PATH, Some("/api/v3/statuses/7/extra")),
            None
        );
    }

    #[test]
    fn a_principal_id_comes_from_any_of_its_three_collections() {
        assert_eq!(
            parse_principal_id_from_href(Some("/api/v3/users/12")),
            Some(12)
        );
        assert_eq!(
            parse_principal_id_from_href(Some("/api/v3/groups/12")),
            Some(12)
        );
        assert_eq!(
            parse_principal_id_from_href(Some("/api/v3/placeholder_users/12")),
            Some(12)
        );
        assert_eq!(
            parse_principal_id_from_href(Some("/api/v3/statuses/12")),
            None
        );
    }
}
